### ISSUE TITLE
User cannot log out

### ISSUE THREAD

**Issue by jonathaneckman** [AzureAD/microsoft-authentication-library-for-js#735]
## I'm submitting a...
- [x] Bug report

## Browser:
- [x] Chrome version XX
- [x] Edge version 76.0.167.1

## Library version
Library version: 1.0.1

## Current behavior
1. Use clicks logout. App uses MSAL logout() to redirect to:

```
https://login.microsoftonline.com/common/oauth2/v2.0/logout
?post_logout_redirect_uri=http://localhost:3000
```

Sometimes, AD stops at this page and asks the user to select which account they should log out of. Other times, it blows right past it and automatically redirects to the next page. 

2. Page redirects (on AAD/B2C side) to:

```
https://login.microsoftonline.com/common/oauth2/v2.0/logoutsession
```

Notice the `post_logout_redirect_uri` WAS in the first page, but is now gone. MSAL seems to be doing its part, but something gets lost during this redirect on the identity provider.

![image](https://user-images.githubusercontent.com/11893200/58671685-7d37c380-8311-11e9-8838-e50d9f2aa378.png)

The page stays there with the message "You signed out of your account. It's a good idea to close all browser windows". **The user has no way to return to the app**.

4. Manually browse back to the app.

5. User does not appear to be logged in. A request is sent to:

```
https://login.microsoftonline.com/common/oauth2/v2.0/authorize
?response_type=id_token%20token 
&scope=profile openid offline_access user_impersonation
&client_id=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXX
&redirect_uri=http%3A%2F%2Flocalhost%3A3000
&state=d2820c44-3e2c-499b-8ebe-7e338c148413
&nonce=0e6504d0-0a8a-4c71-a06a-7143fee1cc58
&client_info=1&x-client-SKU=MSAL.JS
&x-client-Ver=1.0.0
&client-request-id=26def4cb-74bf-4211-8f39-15a4f13ca05f
&prompt=none
```

Which returns with the following from the hidden iFrame injected by MSAL:
"Loading frame has timed out after: 6 seconds for scope profile openid offline_acces"

And the following from a manual call to MSAL acquireTokenSilent():
"Token renewal operation failed due to timeout."

This is okay. I'd expect these to fail.

6. User clicks login button. App calls MSAL loginRedirect(). Page redirects to:

```
https://login.microsoftonline.com/te/mytenant.onmicrosoft.com/b2c_1_signup-signin/oauth2/v2.0/authorize
?response_type=token    
&scope=profile openid offline_access user_impersonation
&client_id=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXX
&redirect_uri=http%3A%2F%2Flocalhost%3A3000
&state=8abd9bb6-605e-4ab3-8f61-0a1ff2e464a7
&nonce=35d90353-9a84-4099-90e2-2ce49a272994
&client_info=1&x-client-SKU=MSAL.JS
&x-client-Ver=1.0.0
&login_req=ba5c17b8-ad9a-4570-89b7-7edc88e11eb2-b2c_1_signup-signin
&domain_req=b39b039b-fe33-4653-8e63-7a7f59b3c864
&domain_hint=organizations
&client-request-id=cbe3310c-3ef5-460f-aae9-fa47973b5461
&prompt=none
```

**User is NOT promped for credentials. They are authenticated and automatically redirected back to the app.** 

## Expected behavior
1. I expect the user to not have to select what account they want to log out of when they are only logged in with 1 account.
1. I'd expect them to get redirected back to the app.
1. I'd expect them to have to provide username and password after they appear to have been logged out in already.

## Other

- I am working with Azure support on this as well, but so far they have seen nothing incorrectly configured on the B2C or Azure end. I decided to post it here in case there is something library related that will help.
- If the user closes and reopens the browser, they are fully logged out. 
- Before clicking Login and skipping the credentials, there are no cookies or any info stored locally. I have no idea how it is bypassing the credential prompt.

**Comment by DarylThayil** [comment]
@jonathaneckman  Thank you for the comment.  We do know there are some scenarios where the server is currently not always redirecting back to webpage.  This is something we are working with that team on.

There are different storage options, sessionStorage which will hold true to the browser session, and localStorage which will persist longer. 


@sameerag  / @pkanher617  do you have any extra context to add?

**Comment by jonathaneckman** [comment]
Thanks @DarylThayil. I do see the same behavior with both storage options. Whats strange is there there is nothing in local or session storage at all when the user gets logged back in again without a credential prompt. I have no idea how its logging me back in. Best guess is its cached in the iframe and I cant see it in developer tools. 

![image](https://user-images.githubusercontent.com/11893200/59159867-aab50780-8a9d-11e9-9e46-1638e9219549.png)

Unfortunately support said its not on the back end so there must be something wrong with the app code. They suggested the logout end point is incorrect, which is set internally by the MSAL library. I'd try other endpoints, but dont see any other options documented. Support was only able to send [this page](https://docs.microsoft.com/en-us/azure/active-directory-b2c/b2clogin) which speaks only to login flow, not logout.

This is a dead end for us so we'll have to look toward other options, but I'll keep this code around in case  there are updates you'd like us to try.

**Comment by DarylThayil** [comment]
So just to be clear when a user logs in, the cache is empty? but you still have access to the account?

**Comment by verbedr** [comment]
@jonathaneckman Is it possible to share how you have configured msal.  I have just recently set up a B2C with msal@1.0.1 (in a test setup with Aurelia framework). And I only configured the authority and client id and it all works as it should.

What is strange in your explanation is that the logout url is not in line with the login url.  If you specified the authority can you check the <authority>/v2.0/.well-known/openid-configuration and check if the authorization, token and end_session endpoints are all in line.

Sinse v1.0.1 the end_session_endpoint is used to log out. My authority looks like `https://login.microsoftonline.com/te/<tenant>.onmicrosoft.com/<policy>/`. And it will logout using `https://login.microsoftonline.com/te/<tenant>.onmicrosoft.com/<policy>/oauth2/v2.0/logout`

I'm not that involved in this project, but I hope this information could be useful to your situation.

**Comment by jonathaneckman** [comment]
@DarylThayil that is correct. 
1. Cache initially empty.
1. User logs in. Redirected to AAD username/password form. Redirected back to app.
1. Values added to cache.
1. User logs out. Redirected to AAD signout page. Never redirected back.
1. Cache is empty.
1. Click login button again. Redirected to AAD. Username/password form flashes. Redirects back to app without credentials. 
1. Values are in the cache.  

@verbedr  Here is my configuration:

```
FLOW_SIGN_IN_SIGN_UP = "B2C_1_signup-signin";
authority: string = "https://mytenant.b2clogin.com/mytenant.onmicrosoft.com";

let userAgentConfig: Configuration = {
    auth: {
        clientId: environmentVariables.appsettings.auth.clientId,
        redirectUri: 'https://localhost:3000',
        postLogoutRedirectUri: 'https://localhost:3000', 
        validateAuthority: false
    },
    cache: {
        cacheLocation: 'localStorage'
    },
    system: {
        logger: this.logger
    }
}

this.userAgentApp = new UserAgentApplication(userAgentConfig);

login() {
    let authParams: AuthenticationParameters = {
        scopes: [
            "openid", 
            "profile"
        ]
    };
    this.userAgentApp.authority = `${this.authority}/${this.FLOW_SIGN_IN_SIGN_UP}`;
    this.userAgentApp.loginRedirect(authParams);
}

logout() {
    this.userAgentApp.logout();
}
```

**Comment by jonathaneckman** [comment]
I published the recorded steps so you can see:

[https://youtu.be/XIiQZL7pERQ](https://youtu.be/XIiQZL7pERQ)

**Comment by DarylThayil** [comment]
This is really helpful, will share with our server team and see if we can get to bottom. I think that what is happening is that the cookies for login.microsoft are not being cleared

**Comment by DarylThayil** [comment]
Okay, talked to our server team, I think the issue is that this is b2c authority, but we are logging out from our aad v2 endpoint 
@sameerag  this should be using the logout url in the app metadata correct?

**Comment by panduit-arg** [comment]
hey @DarylThayil is there any update on this issue? I'm running into the same issue of users not being able to logout and not being redirected after a logout.

**Comment by PraveenVerma17** [comment]
Hi Guys,

We show list of user during logout to select account even though we have one account. 

I have end session endpoint configured properly and we are using msal 1.0.1

"end_session_endpoint":"https://login.microsoftonline.com/74352aa5-d883-4e4a-8423-7862b342708e/oauth2/v2.0/logout"

We are promoting user list to select account during logout. That’s really annoying. 

This does not happen with ad v1. 

Is there any plan to fix this?

**Comment by DarylThayil** [comment]
definitely want to resolve, trying to understand full scope @jasonnutter  can you follow up on this one?

**Comment by sameerag** [comment]
@DarylThayil This is by design by the AAD service currently on V2. More details in #490; This V2 behavior is added for additional security by the service team. 

Brief context from the server team:
In the security considerations of the OpenId Session Mgmt spec:
The id_token_hint parameter to a logout request can be used to determine which RP initiated the logout request. Logout requests without a valid id_token_hint value are a potential means of denial of service; therefore, OPs may want to require explicit user confirmation before acting upon them. Hence the current behavior is by design, even though it is an extra prompt compared to v1. 

There were two user experience considerations leading to behavior changes in v2:
1) On the v1 protocol, apps logging out users contrary to user expectation are somewhat common. The service wanted to follow some form of the standard’s recommendation to improve the user experience on v2 as compared to v1.
2) On the v1 protocol, most users have a single (AAD) account. On v2, many users have multiple (AAD+MSA) accounts. We feel that sign-out should show accounts that are still signed in after sign-out. Some interpretations of the regulatory requirements also point in this direction.

With these considerations in mind, it was decided that the initial release of the v2 logout should always prompt the user for which account to use. 

We will keep you appraised on when this feature is in scope for server team but for now, the behavior may stay for a while.

**Comment by PraveenVerma17** [comment]
Thanks @sameerag @DarylThayil 
In thread 490 its mentioned that feature will come after a quarter in v2.0. Will wait and monitor these threads.    
Thanks for your support.

**Comment by jfbloom22** [comment]
I am running into the same issue.  What I found to be particularly painful for users is if they choose the wrong account to login with.  They choose an account that has not signed up yet.  They receive an error informing them as much, click the Login button intending to try and login with another account, but they immediately receive the error again.  No opportunity to choose a different account.  I tried adding the logout() function in the error catch, but this does not help.  I hope this is resolved soon.

**Comment by jfbloom22** [comment]
Found a half decent workaround:  Redirect the user to: https://login.microsoftonline.com/common/oauth2/logout 

Additionally, you can setup a logout redirect back to your app:  https://stackoverflow.com/questions/45935305/azure-ad-logout-url-not-redirecting

**Comment by PraveenVerma17** [comment]
@sameerag 
I think we should priorities this issue because its a huge security whole in process. 

Step 1:  Click on log out
Step 2:  Does not select any account to logout
Step 3: leave system if someone else uses same session window or same browser and refreshes the page 
Step 4: MSAL does not remove this from server because token did not expire yet
Step 5: ANother user get the access of previous user data even though we are explicitly clearing cache and storage

**Comment by DarylThayil** [comment]
Note: Our AAD server team has this in backlog and we will try to get prioritized. 
As soon as the protocol work is done, we will confirm it is fixed in library.
Currently unscheduled.

**Comment by choclaudio** [comment]
I am running into the same issue. I have been scratching my head over this and I can confirm that this is  not on the msal code. I will follow this thread with high hopes and expectations :)

**Comment by choclaudio** [comment]
> Found a half decent workaround: Redirect the user to: https://login.microsoftonline.com/common/oauth2/logout
> 
> Additionally, you can setup a logout redirect back to your app: https://stackoverflow.com/questions/45935305/azure-ad-logout-url-not-redirecting

I can confirm that the workaround works perfectly.

**Comment by dluc** [comment]
Azure portal doesn't allow to use `http` + `localhost` for Logout URLs, not sure if that's a bug or a security restriction, but it might be relevant to the conversation. 

![image](https://user-images.githubusercontent.com/371009/62568367-e07a3080-b841-11e9-86c8-95311ed53647.png)

![image](https://user-images.githubusercontent.com/371009/62568496-2800bc80-b842-11e9-8ac8-05c41c864d9f.png)

![image](https://user-images.githubusercontent.com/371009/62568513-36e76f00-b842-11e9-9639-cd495258e6c8.png)

**Comment by PraveenVerma17** [comment]
Are you using AD v1 or v2? Because I did not see logout url configuration in v2.

**Comment by choclaudio** [comment]
> Azure portal doesn't allow to use `http` + `localhost` for Logout URLs, not sure if that's a bug or a security restriction, but it might be relevant to the conversation.
> 
> ![image](https://user-images.githubusercontent.com/371009/62568367-e07a3080-b841-11e9-86c8-95311ed53647.png)
> 
> ![image](https://user-images.githubusercontent.com/371009/62568496-2800bc80-b842-11e9-8ac8-05c41c864d9f.png)
> 
> ![image](https://user-images.githubusercontent.com/371009/62568513-36e76f00-b842-11e9-9639-cd495258e6c8.png)

It is actually a typo. This should work. your image says locahost. should be localhost. :)

**Comment by dluc** [comment]
sorry about the typo, but `localhost` is not accepted - just a different error message:

![image](https://user-images.githubusercontent.com/371009/62643292-7ecacc80-b8fc-11e9-86db-92d6961f23e3.png)

**Comment by dluc** [comment]
sorry about the typo, but localhost is not accepted - just a different error message:

![image](https://user-images.githubusercontent.com/371009/62643446-d6693800-b8fc-11e9-9b43-7df74a313090.png)

The error occurs even when switching to the new experience, which I think is still buggy

![image](https://user-images.githubusercontent.com/371009/62643569-13cdc580-b8fd-11e9-8d1e-689d84e1d061.png)

**Comment by choclaudio** [comment]
Actually, maybe something they have fixed recently because it would not make sense to allow localhost. I initially made the same mistake but azure did not complain about it.

The LOGOUT URL is for single sign OUT. It is not the callback url when  you log out, that you just add to your redirect URIs. Single sign out is when you log out from another application and azure identity service calls that URL directly to give the application the chance to clear it's local state. Obviously localhost would not solve anything in that scenario.

Sorry, could have probably figured that out and written it to you the  first time, ended up just being happy about finding the typo that I did not think further :)

**Comment by dluc** [comment]
Actually "localhost" is a very common and important scenario for developers.
One needs the OAuth flows to redirect to the app running locally, e.g. http://localhost:3000 when working with ReactJS. "Sign in" needs to allow users coming from localhost and needs to redirect to localhost,  "logout" needs to localhost after signing out, etc.

**Comment by choclaudio** [comment]
It does support localhost, in the redirect uri section. not for the LOGOUT url. The LOGOUT url is another feature. Imagine a user login out from another application. With single sign out the azure services need a way to tap into other applications and say, hey clear this user. That is what that configuration is for.

**Comment by choclaudio** [comment]
What I am trying to say is that you should ignore the logout url field (it is for single sign OUT). It won't work with an SPA alone.

You just provide your callback url in the redirect uri section and it will work as you expect it to work with localhost.

**Comment by dluc** [comment]
The logout URL is not a callback URL called by Azure, it's a redirection URL after the logout is complete.

**Comment by dluc** [comment]
> It won't work with an SPA alone.

sorry but this is not correct, I'm working on an SPA right now, and it works as expected.

**Comment by choclaudio** [comment]
Maybe I am missing something with iframes and how this works.

But I have a working setup where I call : https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=http://localhost:4200/logout" and just adding this to the  redirect uri section. The single sign out field is empty for me. The logout works fine, but I am not expecting a single sign out experience, since there is no way for the identity service to call OTHER applications to clear their state.

**Comment by dluc** [comment]
The issue I'm having with the inability of using localhost in the logout URL, is that the logout process breaks during local development, but then it works in production. As a workaround I've created a short URL, using HTTPS, just for the localhost scenario, but this is just a hacked workaround. In prod, the logout url brings the user back to my SPA as expected, but *only if* I configure it in the Azure portal. Anyway, this is what I found out so far after spending some days and after wasting sometimes due to the difference between localhost and prod env.

The way I'm using the auth flow is all based on the user browser following URLs configured in AAD, so for instance when I call the MSAL library signout method, MSAL internally deletes its own data (ie. it doesn't expect Azure to do that), then MSAL redirects the user browser to AAD website (ie different domain) which asks the browser to clear AAD data/cookies, and AAD backend on its own clears some cache about the identity being active. 

When this is all done, AAD tells the browser to go to the logout URI, which in my case is just a way to bring the user back to the app (which for instance is critical for SPAs opened inside a mobile app). When coming back to the app, MSAL has already deleted its data, so from an app perspective the app is signed out. 

However, without this round-trip to sign the user out from MSA, the user (for example) would be unable to log in with a different identity.

**Comment by choclaudio** [comment]
Btw, I just tried to insert http://localhost in my logout url and it worked just fine. I am using the v2 API.

**Comment by choclaudio** [comment]
I ended up not using msal because of this particular bug (this thread). So I rolled my own and followed the specs, including checking the signature against jwks key on browsers that support windows.crypto.subtle (something that I do not think msal.js is doing?). I am using the redirect flow, so no hidden iframes (this is what I guess would be used for single sign out if its not a call from the server as I originally thought). In my particular case I can't use the logout URL field since I have no iframe. But by just logging out using https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=http://localhost:4200/logout and clearing my session and local storage all works fine, including the redirect after a logout.

**Comment by choclaudio** [comment]
Hey, read this article. It is recent:
https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-protocols-openid-connect-code#send-a-sign-out-request

It suggests that my initial thoughts are correct regarding the behaviour of single sign out.

**Comment by whitwa** [comment]
Hi,

We are attempting to use the workaround of redirecting to the following URL:

https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=https%3a%2f%2fmyurl%2fapp%2fsecurity%2flogin.aspx

It is working correctly for Work accounts but not for any Personal accounts. It just sticks at the "You signed out of your account - it's a good idea to close all browser windows" screen instead of redirecting to our login page and doesn't actually logout.

Any ideas on how we can get this to work with Personal accounts?

Thanks in advance.

**Comment by amnporter** [comment]
> Hi,
> 
> We are attempting to use the workaround of redirecting to the following URL:
> 
> https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=https%3a%2f%2fmyurl%2fapp%2fsecurity%2flogin.aspx
> 
> It is working correctly for Work accounts but not for any Personal accounts. It just sticks at the "You signed out of your account - it's a good idea to close all browser windows" screen instead of redirecting to our login page and doesn't actually logout.
> 
> Any ideas on how we can get this to work with Personal accounts?
> 
> Thanks in advance.

We see the same thing when dealing with office365 vs outlook.com email accounts.

**Comment by jasonnutter** [comment]
@whitwa @amnporter The logout screen not redirecting back to your app after logout for MSA accounts is a known issue, unfortunately.

**Comment by whitwa** [comment]
Thanks @jasonnutter

What are the timescales for this issue to be resolved? This is preventing us from going live using Microsoft Account Login.

**Comment by jasonnutter** [comment]
@whitwa Unfortunately, I don't have an ETA for when it will be fixed, as it's a bug on login.microsoft.com, which is owned by another team. We'll try to find more information from them, and share back if we hear something definitive. Sorry about that :/

**Comment by DarylThayil** [comment]
@hpsin can you provide context on this

**Comment by hpsin** [comment]
We have a livesite opened on the redirect issue on sign-in, and I expect it may be related to the sign-out behavior you're seeing.  

V2 does not support UX-less signout at this time due to Denial of Service attacks inherent in the spec (and warned of in the spec).

**Comment by rmagor** [comment]
Are there any updates on this issue? This is preventing us from implementing an idle timeout (where the user is automatically logged out when they are inactive for too long). Would prefer to redirect them to a "You were inactive and logged out" page in our app rather than the current "Which account would you like to sign out of page"

**Comment by jasonnutter** [comment]
@rmagor As noted by @hpsin, we do not support silent sign out in order to prevent DDOS attaches.

**Comment by sameerag** [comment]
@sameerag  to internally follow up with the STS team to track this.
@hpsin Can you help us understand if this will be followed up from the STS in the future to support silent logout?

**Comment by hpsin** [comment]
We need to get the OIDC/oauth specs updated to provide a better mechanism here. Until that happens we don't have plans to improve the logout flow - it will remain interactive.

**Comment by sameerag** [comment]
Thanks @hpsin for clarifying this. Closing the issue as there is not much msal js can do here.

**Comment by amnporter** [comment]
So if we popup the logout window and the user logs out, is there a callback that can be received to the webapp that logout was successful?

**Comment by amnporter** [comment]
> > Hi,
> > We are attempting to use the workaround of redirecting to the following URL:
> > https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=https%3a%2f%2fmyurl%2fapp%2fsecurity%2flogin.aspx
> > It is working correctly for Work accounts but not for any Personal accounts. It just sticks at the "You signed out of your account - it's a good idea to close all browser windows" screen instead of redirecting to our login page and doesn't actually logout.
> > Any ideas on how we can get this to work with Personal accounts?
> > Thanks in advance.
> 
> We see the same thing when dealing with office365 vs outlook.com email accounts.
> 
> In addition, is there a callback for when the logout is completed? We're using `https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=xxxxxxxxx`. Currently we leave the onus on the user, but the app doesn't know if they actually logged out of MS.

Any response to the callback here? We're not seeing that the webapp will know that the user actually logged out.

**Comment by BoshJeckerleg** [comment]
For anyone who is stumped on this and finds this thread, I found a solution. 

Similar to how you have to explicitly set your clientApplication authority before you redirect to login

`    this.clientApplication.authority = https://${msalTenant}.b2clogin.com/${msalTenant}.onmicrosoft.com/${msalSignInPolicy};
    this.clientApplication.loginRedirect();`

You have to explicitly set this before triggering logout, otherwise you are directed to the common MS online logout page. 

`    this.clientApplication.authority = https://${msalTenant}.b2clogin.com/${msalTenant}.onmicrosoft.com/${msalSignInPolicy};
    this.clientApplication.logout();`

I figured this out by examining the network call to `openid-configuration` when the login process is triggered. This comes back with an object that has the property `end_session_endpoint` which matches the URL pattern above. 

In the source code for logout it looks for an end session endpoint and if it doesnt exist, redirects you to common login online for logout. Where your redirect URI is not configured and it breaks. 

tldr;
Set the client application authority before logging out. 
Set the logout redirect uri as a reply URL in your azure B2C application configuration.

**Comment by DarylThayil** [comment]
@BoshJeckerleg  thanks for sharing the solution! Login redirect takes a request object where you can define authority, but looks like logout may want this too
cc @sameerag  for thoughts.

**Comment by johanovic** [comment]
Is this related to the following?

If you have multiple accounts that all have the same permissions for the same app, the login will automatically select one when you try to add another account. You don't get the time to log in a different account.

**Comment by m-sala-digital88** [comment]
Hi guys, i've just subrscribed in github cos i got the same problem of @jonathaneckman  and i've just found out the solution of this problem. When you call your method to logout the user you've to set the  authority field again cos you setted it only for the login method and the logout method doesn't know how and who logging out. 
I think changing your logout method like this could work fine :-)

`logout() {
    this.userAgentApp.authority = `${this.authority}/${this.FLOW_SIGN_IN_SIGN_UP}`;
    this.userAgentApp.logout();
}`

let me know. See you!

**Comment by jfbloom22** [comment]
I recently updated to 1.2.1 and this is still an issue.  
**Issue summary:** 
`msalApp.logout()` appears to log the user out, but after "logging out" the user can click "login" and be logged in again without being required to enter their username and password.

This is a serious security issue for user's who login to our application on a public computer, then logout thinking that they have prevented someone from accessing their account.  

**potential solution: re-setting the authority**
I tried re-setting the authority like @m-sala-digital88 and @BoshJeckerleg suggested, but clicking the login button still logs the user in without prompting for a username and password.  

**workarounds**
1. Redirect the user to: https://login.microsoftonline.com/common/oauth2/logout
2. Redirect the user to: https://login.microsoftonline.com/<tenantid>/oauth2/logout?client_id=<clientId>&post_logout_redirect_uri=<encodedurl>
3. Redirect the user to: https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=xxxxxxxxx

The issue with 2 and 3 in my situation, is that we have two web applications sharing this Azure Active Directory and the Azure portal only accepts a single `logout url`.  

The other issue with with this workaround is that in my case we also allow users to login with username and password.  For users who opt to login with username and password rather than their microsoft live account,  redirecting them to a logout URL does not log them out.

@jasonnutter and @sameerag I see that you closed this issue because `we do not support silent sign out in order to prevent DDOS attaches.` .  Obviously, this thread morphed into something different than the original issue.  Would it be better if I open a new issue?  In my summary I did not include some of the issues the original post included.  ```I expect the user to not have to select what account they want to log out of when they are only logged in with 1 account.
I'd expect them to get redirected back to the app.```.   and because of this I hope my summary clarifies and prioritizes the main issue.

**Comment by hpsin** [comment]
@jfbloom22 can you send me a fiddler trace of this behavior? hirsin, microsoft.com

**Comment by jfbloom22** [comment]
hey @hpsin thank you for looking into this.  I am trying to capture a fiddler trace, but I have never used the program before. The Mac version does not seam to work at all.  Wireshark is working, but it will take me some time to figure out how to find and share the right packets.  Before I attempt to do that here is what I see in Chrome Dev tools:

There are two Logout requests:
```
Request URL: https://login.microsoftonline.com/te/7bec88f8-013e-403f-8cc9-9e451c46f7ea/b2c_1_mmg_test_signin/oauth2/v2.0/logout?post_logout_redirect_uri=http%3A%2F%2Flocalhost%3A3000
```

```
Request URL: https://login.microsoftonline.com/556e6b1f-b49d-4278-8baf-db06eeefc8e9/oauth2/v2.0/logout
```
Followed by a login request (where it does Not require the user to re-enter their password)
```
Request URL: https://login.microsoftonline.com/te/7bec88f8-013e-403f-8cc9-9e451c46f7ea/b2c_1_mmg_test_signin/oauth2/v2.0/authorize?response_type=id_token&scope=https%3A%2F%2FInteractionPlacesB2C1.onmicrosoft.com%2Fmmg-api%2Fread%20openid%20profile&client_id=6fa0ff31-bfad-4d6a-99d2-3f7d00cc0b2a&redirect_uri=http%3A%2F%2Flocalhost%3A3000&state=f2c3e9f5-0235-4736-b9f0-839207ae2c02&nonce=a85c038b-6393-48f5-b2bb-2c33828215af&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=1.2.1&client-request-id=8c49ac57-b32f-448e-86fb-3a3e8c6e8220&response_mode=fragment
```

Does that give you enough information?

**Comment by hpsin** [comment]
Ah, yeah, mostly that it's a b2c concern that I can't help with I'm afraid. Apologies, I didn't realize that. @jasonnutter can you pull in someone from b2c instead?

**Comment by jasonnutter** [comment]
@ktsakas Can you provide any insight for this issue? Thanks!

**Comment by bojantv** [comment]
Hi, sorry for getting in the middle of your conversation. I have also noticed issues with logout and session not getting destroyed. However, it happens when the single sign-on configuration is set on "Application". When on "Tenant" it works as expected.

Expected behavior for Application setting is:
When you log out of your app, the session is destroyed and when you visit the same protected page you are redirected to the login screen.
Actual:
When you log out of your app, the session stays active and when you get back to your protected page it silently logs you in. 

On Tenant config the session stays alive till all "tenant users" don't logout.

![image](https://user-images.githubusercontent.com/58706872/72829022-cf00d580-3c75-11ea-9e9c-5371a1bb5bb1.png)

hope it helps, and you get that resolved. 

Thanks

**Comment by jasonnutter** [comment]
@bojantv Can you please open a new issue, thanks!

**Comment by Glideh** [comment]
I'm neither able to logout a user.
I have the exact same behavior as the OP.
Should I create a new issue with the same title and content ?

**Comment by hannonjohn** [comment]
> For anyone who is stumped on this and finds this thread, I found a solution.
> 
> Similar to how you have to explicitly set your clientApplication authority before you redirect to login
> 
> ` this.clientApplication.authority = https://${msalTenant}.b2clogin.com/${msalTenant}.onmicrosoft.com/${msalSignInPolicy}; this.clientApplication.loginRedirect();`
> 
> You have to explicitly set this before triggering logout, otherwise you are directed to the common MS online logout page.
> 
> ` this.clientApplication.authority = https://${msalTenant}.b2clogin.com/${msalTenant}.onmicrosoft.com/${msalSignInPolicy}; this.clientApplication.logout();`
> 
> I figured this out by examining the network call to `openid-configuration` when the login process is triggered. This comes back with an object that has the property `end_session_endpoint` which matches the URL pattern above.
> 
> In the source code for logout it looks for an end session endpoint and if it doesnt exist, redirects you to common login online for logout. Where your redirect URI is not configured and it breaks.
> 
> tldr;
> Set the client application authority before logging out.
> Set the logout redirect uri as a reply URL in your azure B2C application configuration.

Thanks very much @BoshJeckerleg - this sorted my issue.
